### PR TITLE
feat: basic-auth support and minimal login page

### LINKED ISSUE #395: password required when logging in
### Description

hi 👋, I just started trying marimo.

I deployed this as a public service on the server, considering that there might be some security risks. I'm hoping to restrict logins in some way.

Not sure if I've missed something, if marimo can do this please point it out.

Thank you for your great work!

### Suggested solution

If a password is set in the configuration, you need to provide the password to enter marimo.

### Alternative

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mscolnick**
Fixes #395

This adds basic auth support. It is both _basic_ and uses the `Basic` auth schema.

Overview:
* it is enabled by default for `marimo edit/tutorial/new`, and not enabled by default for `marimo run`
* You can enable/disable with `--token/--no-token` and `--token-password` to customize the password
* Rename `ValidateServerTokenMiddleware` to `SkewProtectionMiddleware`

---

In order to authenticate, you must either pass the token as a password in the `Authorization` header, or as a query parameter under `access_token` in the URL.

1. Basic Authorization header:

To authenticate using the `Authorization` header, you must pass the token as a password in the `Authorization` header using the [Basic authentication scheme](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication). For example, to authenticate with the token `sup3rs3cr3t`, you would pass the header `Authorization Basic base64("any_username:sup3rs3cr3t")`.

This is not necessary when using a browser, as we will redirect you to a minimal login page

2. Query parameter (most common flow):

To authenticate using a query parameter, you must pass the token as a query parameter under `access_token` in the URL. For example, to authenticate with the token `sup3rs3cr3t`, you would pass the query parameter `http://localhost:2718?access_token=sup3rs3cr3t`.

For convenience, when running locally, marimo will automatically open the URL with the query parameter in your default browser.

### CODE REVIEW

File: docs/guides/authentication.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# Authentication
+
+marimo provides a simple way to add token/password protection to your marimo server. Given that authentication is a complex topic, marimo does not provide a built-in authentication/authorization system, but instead makes it easy to add your own through ASGI middleware.
+
+## Enabling Basic Authentication
+
+Authentication is disabled by default, however this may change in the future. To enable authentication, you must pass `--token` to your `marimo edit/run/new` command from the Terminal. This will use a randomly generated token in `Edit mode` and a deterministic token in `Run mode` (based on the code of the notebook).
```
Comment (akshayka) [inline docs/guides/authentication.md:7]: @mscolnick Update this sentence to indicate that auth is enabled by default in `edit/tutorial/new` ?

File: marimo/_server/api/auth.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        self, scope: Scope, receive: Receive, send: Send
+    ) -> None:
+        if scope["type"] not in ("http", "websocket"):
+            await self.app(scope, receive, send)
+            return
+
+        state = AppState.from_app(scope["app"])
+        # disabled = state.session_manager.mode != SessionMode.EDIT
+        # if disabled:
+        #     return await self.app(scope, receive, send)
```
Comment (akshayka) [inline marimo/_server/api/auth.py:180]: Delete dead code?

File: marimo/_server/tokens.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    @staticmethod
+    def from_code(code: str) -> SkewProtectionToken:
+        return SkewProtectionToken(str(hash(code)))
+
+    @staticmethod
+    def random() -> SkewProtectionToken:
+        return SkewProtectionToken(secrets.token_urlsafe(16))
+
+    def __str__(self) -> str:
+        return self.token
```
Comment (akshayka) [inline marimo/_server/tokens.py:44]: nice abstraction 👌

File: pyproject.toml [hunk 4]
```diff
@@ -39,6 +39,8 @@ dependencies = [
     "docutils>=0.17.0",
     # to show RAM, CPU usage, other system utilities
     "psutil>=5.0",
+    # required dependency in Starlette for SessionMiddleware support
+    "itsdangerous>=2.0.0",
```
Comment (akshayka) [inline pyproject.toml:43]: 👍

**Review by akshayka — COMMENTED** [review]
Nice work --  very thorough. Some review comments left in-line. Additionally:

1. Can we have a more intuitive token/auth landing page? I got a browser alert with a username and password, which may confuse our users.

image

Instead, how about just a single field that asks for the token?

2. Why `access_token` instead of just `token` as a query param? Is it to distinguish among multiple types of tokens? I like just `token` but it's fine if there's a need to disambiguate.

**Comment by mscolnick** [comment]
minimal login page 

image

**Review by akshayka — APPROVED** [review]
LGTM!

@dmadisetti, want to review?

**Comment by dmadisetti** [comment]
This looks awesome. I might do a bit of a security audit for fun later !

**Pull request merged**